File: research/experiments/sweep/runs/opt-full-csharp-sonnet45/src/Bfgs.cs
namespace Optimization;

public static class Bfgs
{
    public static OptimizeResult Minimize(
        Func<double[], double> f, double[] x0,
        Func<double[], double[]>? grad = null, OptimizeOptions? options = null)
    {
        var opts = ResultTypes.MergeOptions(options);
        var gradFn = grad ?? (x => FiniteDiff.ForwardDiffGradient(f, x));
        int n = x0.Length;

        double[] x = (double[])x0.Clone();
        double fx = f(x);
        double[] gx = gradFn(x);
        int functionCalls = 1;
        int gradientCalls = 1;

        double gradNorm = VecOps.NormInf(gx);
        if (gradNorm <= opts.GradTol)
        {
            return new OptimizeResult
            {
                X = x, Fun = fx, Gradient = (double[])gx.Clone(),
                Iterations = 0, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                Converged = true, Message = "Convergence: gradient norm below tolerance"
            };
        }

        // Initialize inverse Hessian as identity
        double[][] H = IdentityMatrix(n);

        for (int iter = 1; iter <= opts.MaxIterations; iter++)
        {
            // Direction d = -H * g
            double[] d = MatVecMul(H, VecOps.Negate(gx));

            var ls = LineSearch.WolfeLineSearch(f, gradFn, x, d, fx, gx);
            functionCalls += ls.FunctionCalls;
            gradientCalls += ls.GradientCalls;

            if (!ls.Success)
            {
                return new OptimizeResult
                {
                    X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
                    Iterations = iter, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                    Converged = false, Message = "Stopped: line search failed"
                };
            }

            double[] xNew = VecOps.AddScaled(x, d, ls.Alpha);
            double fNew = ls.FNew;
            double[] gNew = ls.GNew ?? gradFn(xNew);
            if (ls.GNew == null) gradientCalls++;

            double[] sk = VecOps.Sub(xNew, x);
            double[] yk = VecOps.Sub(gNew, gx);
            double yts = VecOps.Dot(yk, sk);

            double stepNorm = VecOps.NormInf(sk);
            double funcChange = Math.Abs(fNew - fx);

            x = xNew; fx = fNew; gx = gNew;
            gradNorm = VecOps.NormInf(gx);

            // Curvature guard
            if (yts > 1e-10)
            {
                double rhoK = 1.0 / yts;
                BfgsUpdate(H, sk, yk, rhoK);
            }

            var reason = ResultTypes.CheckConvergence(gradNorm, stepNorm, funcChange, iter, opts);
            if (reason != null)
            {
                return new OptimizeResult
                {
                    X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
                    Iterations = iter, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
                    Converged = ResultTypes.IsConverged(reason), Message = ResultTypes.ConvergenceMessage(reason)
                };
            }
        }

        return new OptimizeResult
        {
            X = (double[])x.Clone(), Fun = fx, Gradient = (double[])gx.Clone(),
            Iterations = opts.MaxIterations, FunctionCalls = functionCalls, GradientCalls = gradientCalls,
            Converged = false, Message = "Stopped: reached maximum iterations"
        };
    }

    public static double[][] IdentityMatrix(int n)
    {
        double[][] I = new double[n][];
        for (int i = 0; i < n; i++)
        {
            I[i] = new double[n];
            I[i][i] = 1.0;
        }
        return I;
    }

    public static double[] MatVecMul(double[][] M, double[] v)
    {
        int n = v.Length;
        double[] result = new double[n];
        for (int i = 0; i < n; i++)
        {
            double sum = 0;
            for (int j = 0; j < n; j++) sum += M[i][j] * v[j];
            result[i] = sum;
        }
        return result;
    }

    public static void BfgsUpdate(double[][] H, double[] s, double[] y, double rho)
    {
        int n = s.Length;
        // H_new = (I - rho*s*y') * H * (I - rho*y*s') + rho*s*s'
        double[] Hy = MatVecMul(H, y);
        double yHy = VecOps.Dot(y, Hy);

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                H[i][j] = H[i][j] - rho * (s[i] * Hy[j] + Hy[i] * s[j]) + rho * (rho * yHy + 1) * s[i] * s[j];
            }
        }
    }
}

File: research/experiments/sweep/runs/opt-medium-python-sonnet45/src/bfgs.py
"""
bfgs — BFGS quasi-Newton optimizer.

Full-memory BFGS with inverse Hessian approximation.
"""

from typing import Callable, Optional
import vec_ops
from result_types import OptimizeOptions, OptimizeResult, default_options, check_convergence, convergence_message
from line_search import wolfe_line_search
from finite_diff import forward_diff_gradient


def identity_matrix(n: int) -> list[list[float]]:
    """Create n×n identity matrix as list of row lists."""
    matrix = []
    for i in range(n):
        row = [0.0] * n
        row[i] = 1.0
        matrix.append(row)
    return matrix


def mat_vec_mul(M: list[list[float]], v: list[float]) -> list[float]:
    """Matrix-vector multiplication (M is list of row lists)."""
    return [vec_ops.dot(row, v) for row in M]


def bfgs_update(
    H: list[list[float]],
    s: list[float],
    y: list[float],
    rho: float
) -> list[list[float]]:
    """
    Apply BFGS inverse Hessian update formula.

    H_{k+1} = (I - rho*s*y^T) * H * (I - rho*y*s^T) + rho*s*s^T
    where rho = 1/(y^T*s)

    Using efficient expanded form:
    H_ij - rho*(s_i*Hy_j + Hy_i*s_j) + rho*(1 + rho*y^T*H*y)*s_i*s_j
    """
    n = len(s)
    Hy = mat_vec_mul(H, y)
    yHy = vec_ops.dot(y, Hy)

    H_new = []
    for i in range(n):
        row = []
        for j in range(n):
            row.append(
                H[i][j] -
                rho * (s[i] * Hy[j] + Hy[i] * s[j]) +
                rho * (1 + rho * yHy) * s[i] * s[j]
            )
        H_new.append(row)

    return H_new


def bfgs(
    f: Callable[[list[float]], float],
    x0: list[float],
    grad: Optional[Callable[[list[float]], list[float]]] = None,
    options: Optional[OptimizeOptions] = None
) -> OptimizeResult:
    """
    Minimize a function using BFGS quasi-Newton method.

    Args:
        f: Objective function
        x0: Initial point
        grad: Gradient function (optional, uses finite differences if not provided)
        options: Optimization options

    Returns:
        OptimizeResult with solution and diagnostics
    """
    opts = default_options(options)

    # Use finite differences if no gradient provided
    grad_fn = grad if grad is not None else lambda x: forward_diff_gradient(f, x)

    n = len(x0)
    x = vec_ops.clone(x0)
    H = identity_matrix(n)

    function_calls = 0
    gradient_calls = 0

    # Initial evaluation
    fx = f(x)
    function_calls += 1

    gx = grad_fn(x)
    gradient_calls += 1

    # Check if already at minimum (use infinity norm)
    grad_norm = vec_ops.norm_inf(gx)
    initial_reason = check_convergence(grad_norm, float('inf'), float('inf'), 0, opts)
    if initial_reason and initial_reason["kind"] in ("gradient", "step", "function"):
        return OptimizeResult(
            x=x,
            fun=fx,
            gradient=gx,
            iterations=0,
            function_calls=function_calls,
            gradient_calls=gradient_calls,
            converged=True,
            message=convergence_message(initial_reason)
        )

    for iteration in range(1, opts["max_iterations"] + 1):
        # Compute search direction: d = -H * g
        d = vec_ops.negate(mat_vec_mul(H, gx))

        # Perform Strong Wolfe line search
        ls_result = wolfe_line_search(f, grad_fn, x, d, fx, gx)
        function_calls += ls_result.function_calls
        gradient_calls += ls_result.gradient_calls

        if not ls_result.success:
            return OptimizeResult(
                x=x,
                fun=fx,
                gradient=gx,
                iterations=iteration,
                function_calls=function_calls,
                gradient_calls=gradient_calls,
                converged=False,
                message="Stopped: line search failed"
            )

        # Update position
        x_new = vec_ops.add_scaled(x, d, ls_result.alpha)
        fx_new = ls_result.f_new
        gx_new = ls_result.g_new if ls_result.g_new is not None else grad_fn(x_new)

        if ls_result.g_new is None:
            gradient_calls += 1

        # Compute step and gradient change
        s = vec_ops.sub(x_new, x)
        y = vec_ops.sub(gx_new, gx)

        # Use infinity norm for convergence checks
        step_norm = vec_ops.norm_inf(s)
        grad_norm = vec_ops.norm_inf(gx_new)
        func_change = abs(fx_new - fx)

        # Update state
        x = x_new
        fx = fx_new
        gx = gx_new

        # Check convergence
        reason = check_convergence(grad_norm, step_norm, func_change, iteration, opts)
        if reason is not None:
            return OptimizeResult(
                x=x,
                fun=fx,
                gradient=gx,
                iterations=iteration,
                function_calls=function_calls,
                gradient_calls=gradient_calls,
                converged=(reason["kind"] in ("gradient", "step", "function")),
                message=convergence_message(reason)
            )

        # Curvature condition check - skip BFGS update if not satisfied
        ys = vec_ops.dot(y, s)
        if ys <= 1e-10:
            continue

        # Update inverse Hessian approximation
        rho = 1.0 / ys
        H = bfgs_update(H, s, y, rho)

    # Max iterations reached - this shouldn't normally be reached
    return OptimizeResult(
        x=x,
        fun=fx,
        gradient=gx,
        iterations=opts["max_iterations"],
        function_calls=function_calls,
        gradient_calls=gradient_calls,
        converged=False,
        message="Stopped: maximum iterations reached"
    )
